lib/ecto/type.ex

```diff
@@ -291,7 +293,20 @@ defmodule Ecto.Type do
     {:ok, %Ecto.Query.Tagged{value: nil, type: type(type)}}
   end
 
-  def dump({:array, type}, value) do
+  def dump({:embed, %{embed: model}}, %{__struct__: model} = struct) do
```

Comment:
I don't check it exactly for recursing in embed_many, I think I need to extract it to a separate function.